src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,147 @@
+    unsigned int num_elements;
+} Trajectory_t;
+
+/*
+ *   t_start - The path parameter value indicating the beginning of the
+ *       considered path
+ *   t_end - The path parameter value indicating the end of the considered
+ *       path
+ *
+ *   num_segments - The number of segments to discretize the trajectory into.
+ *       Must be greater than 2 segments.
+ *       THE NUMBER OF SEGMENTS MUST BE UNDER
+ *       TRAJECTORY_PLANNER_MAX_NUMBER_SEGMENTS
+ *
+ *   max_allowable_acceleration - The maximum acceleration allowed at any
+ *       point along the trajectory. This factor limits the maximum delta-velocity and
+ * also the max speed around curves due to centripetal acceleration [m/s^2]
+ *
+ *   max_allowable_speed - The maximum speed allowable at any point along the
+ *
+ *   path [m/s] *initial_speed - The initial speed at the start of the trajectory [m/s]
+ *
+ *   final_speed - The final speed at the end of the trajectory [m/s]
+ */
+typedef struct FirmwareRobotPathParameters
+{
+    Polynomial2dOrder3_t path;
+    float t_start;
+    float t_end;
+    unsigned int num_segments;
+    float max_allowable_acceleration;
+    float max_allowable_speed;
+    float initial_speed;
+    float final_speed;
+} FirmwareRobotPathParameters_t;
+
+/**
+ * Returns a planned trajectory as a list of TrajectoryElement's.
+ *
+ *  Key assumptions & guarantees of this planner are:
+ *  - Trajectories are time-optimal assuming INFINITE JERK capability of the robot
+ *  - No speed on the trajectory is larger than the 'max_allowable_speed' parameter.
+ *  - No acceleration value between points on the trajectory can be larger than the
+ * 'max_allowable_acceleration' input parameter
+ *  - Assuming the grip-limit of the robot IS THE SAME AS THE MAX ACCELERATION then at no
+ * point on the path can the sum of centripetal and acceleration force be greater than
+ * 'max_allowable_acceleration'
+ *
+ *  Trajectory generation is done by assuming constant acceleration capability.
+ *      - The generator will assume maximum acceleration for the robot between each
+ * segment on the path. If using max acceleration breaches either the speed limit of the
+ * robot, or the centripetal acceleration limit defined by the curvature and robot speed -
+ * the planner will assume the max acceleration that will remain bellow this limit.
+ *
+ *     - To ensure the robot is capable of deccelerating to reach ant sudden changes in
+ * the path, the trajectory is checked for backwards continuity
+ *
+ * @pre num_segments > 2
+ * @pre max_allowable_acceleration > 0
+ * @pre max_allowable_speed > 0
+ * @pre init_speed >= 0
+ * @pre final_speed >= 0
+ *
+ *  @param path_parameters [in] The data structure including important path parameters
+ * such as: *path - The 3D polynomial representing the geometric path
+ */
+void app_trajectory_planner_generate_constant_arc_length_segmentation(
+    FirmwareRobotPathParameters_t path_parameters, Trajectory_t* trajectory);
```

Comment:
Do we need to add a `@pre` that the trajectory is pre-allocated to handle up to the maximum number of elements?

Reply:
Good point, added the @pre